### ISSUE TITLE
"&<letter>" sends OK but does not display  

### ISSUE THREAD

**Issue by 45bcdf** [ubports/messaging-app#332]
Sending a text that finishes with an '&' and a letter will deliver the complete text properly on the receiving device, but in the sender's conversation the text will not display the "&" or the letter after it. 

I have not tested multiple letter variations/combos. The bug was identified by sending "M&M" and only seeing "M" at the end of the sent text. I asked others in the UBports[ENGLISH] telegram channel to confirm they could replicate and at least two others did.

One user seems to have identified: `textFormat: Text.StyleText` as the cause.

**Comment by jezek** [comment]
I'm pretty sure, the #333 fix contains an bug.
Before the `<`, `>` characters were replaced by `&lt;`, `&gt;` to be shown in the message.
After this the tags get erased (and it's causing the #339 issue).
The [tags replace line](https://github.com/ubports/messaging-app/pull/333/files#diff-30c0dde106a01a4abe38741114a00e8c0f7dba9723c8008e88d87254a6833868L67) should be left alone and not be changed.

**Comment by lduboeuf** [comment]
humm, if i let the first step like this, there will be issues with the '&' replacement later. ( &alt will be replaced by "&amp;lt" ) and will cause another issue ^ ^.
You think #339 occured after #333 merged?

**Comment by jezek** [comment]
> You think #339 occured after #333 merged?

Yes. I can check for sure, with going back to the stable and see the message, but I'm pretty sure it is the case.

> humm, if i let the first step like this, there will be issues with the '&' replacement later. ( &alt will be replaced by "&lt" ) and will cause another issue ^ ^.

You're right, didn't realize this. Easy fix. Swap the lines. First escape the `&` and then escape the `<` and `>`. ;)

**Comment by lduboeuf** [comment]
> You're right, didn't realize this. Easy fix. Swap the lines. First escape the `&` and then escape the `<` and `>`. ;)

and what about "&amp or &alt" that may be in the message already

**Comment by jezek** [comment]
> and what about "&amp or &alt" that may be in the message already

Do the messages come in HTML? I think the plaintext messages are not in HTML. So when anybody writes `&amp;` into an message, he wants to be it `&amp;` so it should be translated to `&amp;amp;` to be correctly shown as HTML text.

 I think the whole thing can be collapsed to
```
        // escape '&', '<', '>' chars
        text = text.replace(/&/g, '&amp;').replace(/</g,'&lt;').replace(/>/g,'<tt>&gt;</tt>');
```

If the messages came already in HTML, no escaping would be needed.

Note: I don't know the effect of `<tt></tt>` tags in the last replace, but it was there before, so I left it there.

**Comment by lduboeuf** [comment]
@jezek so you are about to PR something ;-) Would be great. idk though if it will be accepted by the team. But if that current behavior add a regression. Better to revert it ?

**Comment by jezek** [comment]
Was I supposed to PR something?

I just saw that there is an QA announcement in forum, browsed the issues and saw this one, which was similar to the one I wrote. Than I spotted (by looking at the code), that the #333 did the regression and wrote my observations.

> But if that current behavior add a regression. Better to revert it ?

My opinion is to fix it (by leaving the `<`, `>` replace line as it is and moving the `&` replacing above) instead of revert it. But if it is no time for corrections (just for reverts) then yes, I think it is better to revert it (and fix it until next OTA), cause it creates a regression.

**Comment by Flohack74** [comment]
If its a risk-free fix lets do it. But then we need to expedite it a bit, I wanted to release on Fri, so you would need to PR it today or tomorrow latest, so we can run an RC overnight. Due to our infrastructure we need to wait considerable amount of time until we see it in an RC

**Comment by lduboeuf** [comment]
@Flohack74 see https://github.com/ubports/messaging-app/pull/341
did some quick tests, i don't expect any regression

**Comment by jezek** [comment]
I tested the #341 PR of messaging-app on phone using clickable and I must say, that the issue #339 is no longer valid.
Good job.
